tools/lkl/lib/rump-host.c

```diff
@@ -0,0 +1,672 @@
+#include <string.h>
+#include <stdint.h>
+#include <time.h>
+#include <errno.h>
+#include <sys/types.h>
+
+#include <unistd.h>
+#include <poll.h>
+#include <sys/uio.h>
+
+#include <lkl_host.h>
+#include "iomem.h"
+#include "jmp_buf.h"
+#include "rump.h"
+
+
+/* FIXME */
+#define BIT(x) (1ULL << x)
+#define NSEC_PER_SEC	1000000000L
+#define container_of(ptr, type, member) \
+	(type *)((char *)(ptr) - __builtin_offsetof(type, member))
+
+/* FIXME */
+int *__errno(void);
+#undef errno
+#define errno (*__errno())
+
+
+/* console */
+static void rump_print(const char *str, int len)
+{
+	while (len-- > 0) {
+		rumpuser_putchar(*str);
+		str++;
+	}
+}
+
+
+/* semaphore/mutex */
+struct rumpuser_sem {
+	struct rumpuser_mtx *lock;
+	int count;
+	struct rumpuser_cv *cond;
+};
+
+struct lkl_mutex {
+	struct rumpuser_mtx *mutex;
+};
+
+struct lkl_sem {
+	struct rumpuser_sem sem;
+};
+
+static struct lkl_sem *rump_sem_alloc(int count)
+{
+	struct lkl_sem *sem;
+
+	rumpuser_malloc(sizeof(*sem), 0, (void **)&sem);
+	if (!sem)
+		return NULL;
+
+	rumpuser_mutex_init(&sem->sem.lock, RUMPUSER_MTX_SPIN);
+	sem->sem.count = count;
+	rumpuser_cv_init(&sem->sem.cond);
+
+	return sem;
+}
+
+static void rump_sem_free(struct lkl_sem *_sem)
+{
+	struct rumpuser_sem *sem = (struct rumpuser_sem *)&_sem->sem;
+
+	rumpuser_cv_destroy(sem->cond);
+	rumpuser_mutex_destroy(sem->lock);
+	rumpuser_free(sem, 0);
+}
+
+static void rump_sem_up(struct lkl_sem *_sem)
+{
+	struct rumpuser_sem *sem = (struct rumpuser_sem *)&_sem->sem;
+
+	rumpuser_mutex_enter(sem->lock);
+	sem->count++;
+	if (sem->count > 0)
+		rumpuser_cv_signal(sem->cond);
+	rumpuser_mutex_exit(sem->lock);
+}
+
+static void rump_sem_down(struct lkl_sem *_sem)
+{
+	struct rumpuser_sem *sem = (struct rumpuser_sem *)&_sem->sem;
+
+	rumpuser_mutex_enter(sem->lock);
+	while (sem->count <= 0)
+		rumpuser_cv_wait(sem->cond, sem->lock);
+	sem->count--;
+	rumpuser_mutex_exit(sem->lock);
+}
+
+static struct lkl_mutex *rump_mutex_alloc(int recursive)
+{
+	struct lkl_mutex *_mutex;
+
+	rumpuser_malloc(sizeof(*_mutex), 0, (void **)&_mutex);
+	if (!_mutex)
+		return NULL;
+
+	rumpuser_mutex_init(&_mutex->mutex, RUMPUSER_MTX_SPIN);
+
+	return _mutex;
+}
+
+static void rump_mutex_lock(struct lkl_mutex *_mutex)
+{
+	rumpuser_mutex_enter(_mutex->mutex);
+}
+
+static void rump_mutex_unlock(struct lkl_mutex *_mutex)
+{
+	rumpuser_mutex_exit(_mutex->mutex);
+}
+
+static void rump_mutex_free(struct lkl_mutex *_mutex)
+{
+	rumpuser_mutex_destroy(_mutex->mutex);
+	rumpuser_free(_mutex, 0);
+}
+
+/* XXX: dummy TLS */
+struct lkl_tls_key {
+	unsigned int *key;
+};
+static struct lkl_tls_key *rump_tls_alloc(void (*destructor)(void *))
+{
+	return NULL;
+}
+
+static void rump_tls_free(struct lkl_tls_key *key)
+{
+}
+
+static int rump_tls_set(struct lkl_tls_key *key, void *data)
+{
+	rumpuser_curlwpop(RUMPUSER_LWP_SET, (struct lwp *)data);
+	return 0;
+}
+
+static void *rump_tls_get(struct lkl_tls_key *key)
+{
+	return rumpuser_curlwp();
```

Comment:
This doesn't get the TLS variable given the key, correct?  Knowing rumpkernel's handling of TLS (where `get`s are implicit on `__thread` variables), it seems like there might be a mismatch between the LKL and rumpkernel models here.  I like explicit calls to get TLS based on a key in LKL much more.

Reply:
as you can see, this function drops the key argument the result only indicate a TLS val, not localized by a key.  

> I suppose the assumption here is that LKL only uses TLS to store the currently active thread?

Not really, just a limitation of in this RFC stage.